File: core/src/main/cpp/entityauth/PresharedAuthenticationData.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <entityauth/PresharedAuthenticationData.h>
#include <MslEncodingException.h>
#include <MslError.h>
#include <entityauth/EntityAuthenticationScheme.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <io/MslObject.h>
#include <util/MslUtils.h>

using namespace std;
using namespace netflix::msl;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace entityauth {

namespace {
const string KEY_IDENTITY = "identity";
}

PresharedAuthenticationData::PresharedAuthenticationData(const string identity)
    : EntityAuthenticationData(EntityAuthenticationScheme::PSK), identity_(identity)
{
}

// note weird syntax to handle an exception thrown in initializer list
PresharedAuthenticationData::PresharedAuthenticationData(shared_ptr<MslObject> presharedAuthMo)
try
    : EntityAuthenticationData(EntityAuthenticationScheme::PSK)
    , identity_(presharedAuthMo->getString(KEY_IDENTITY))
{
}
catch (const MslEncoderException& e)
{
    stringstream ss;
    ss << "psk authdata " << presharedAuthMo;
    throw MslEncodingException(MslError::MSL_PARSE_ERROR, ss.str(), e);
}


shared_ptr<MslObject> PresharedAuthenticationData::getAuthData(shared_ptr<MslEncoderFactory> factory,
        const MslEncoderFormat& ) const
{
    shared_ptr<MslObject> mo = factory->createObject();
    mo->put<string>(KEY_IDENTITY, identity_);
    return mo;
}

bool PresharedAuthenticationData::equals(std::shared_ptr<const EntityAuthenticationData> obj) const
{
    if (this == obj.get()) return true;
    if (!instanceof<const PresharedAuthenticationData>(obj.get())) return false;
    shared_ptr<const PresharedAuthenticationData> that = dynamic_pointer_cast<const PresharedAuthenticationData>(obj);
    return EntityAuthenticationData::equals(obj) && (identity_ == that->getIdentity());
}

bool operator==(const PresharedAuthenticationData& a, const PresharedAuthenticationData& b)
{
	shared_ptr<const PresharedAuthenticationData> ap(&a, &MslUtils::nullDeleter<PresharedAuthenticationData>);
	shared_ptr<const PresharedAuthenticationData> bp(&b, &MslUtils::nullDeleter<PresharedAuthenticationData>);
	return ap->equals(bp);
}
}}} // namespace netflix::msl::entityauth

File: core/src/main/javascript/entityauth/PresharedAuthenticationData.js
/**
 * Copyright (c) 2012-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>Preshared keys entity authentication data.</p>
 *
 * <p>
 * {@code {
 *   "#mandatory" : [ "identity" ],
 *   "identity" : "string"
 * }} where:
 * <ul>
 * <li>{@code identity} is the entity identity</li>
 * </ul></p>
 *
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
    "use strict";
    
    var EntityAuthenticationData = require('../entityauth/EntityAuthenticationData.js');
    var EntityAuthenticationScheme = require('../entityauth/EntityAuthenticationScheme.js');
    var AsyncExecutor = require('../util/AsyncExecutor.js');
    var MslEncoderException = require('../io/MslEncoderException.js');
    var MslEncodingException = require('../MslEncodingException.js');
    var MslError = require('../MslError.js');
    
    /**
     * Key entity identity.
     * @const
     * @type {string}
     */
    var KEY_IDENTITY = "identity";

    var PresharedAuthenticationData = module.exports = EntityAuthenticationData.extend({
        /**
         * <p>Construct a new preshared keys authentication data instance from the
         * specified entity identity.</p>
         *
         * @param {string} identity the entity identity.
         * @extends {EntityAuthenticationData}
         */
        init: function init(identity) {
            init.base.call(this, EntityAuthenticationScheme.PSK);
            // The properties.
            var props = {
                identity: { value: identity, writable: false }
            };
            Object.defineProperties(this, props);
        },

        /** @inheritDoc */
        getIdentity: function getIdentity() {
            return this.identity;
        },

        /** @inheritDoc */
        getAuthData: function getAuthData(encoder, format, callback) {
            AsyncExecutor(callback, function() {
                var mo = encoder.createObject();
                mo.put(KEY_IDENTITY, this.identity);
                return mo;
            }, this);
        },

        /** @inheritDoc */
        equals: function equals(that) {
            if (this === that) return true;
            if (!(that instanceof PresharedAuthenticationData)) return false;
            return (equals.base.call(this, that) && this.identity == that.identity);
        },
    });

    /**
     * Construct a new preshared keys authentication data instance from the
     * provided MSL object.
     *
     * @param {MslObject} presharedAuthMo the authentication data MSL object.
     * @throws MslEncodingException if there is an error parsing the MSL
     *         representation.
     */
    var PresharedAuthenticationData$parse = function PresharedAuthenticationData$parse(presharedAuthMo) {
        try {
            var identity = presharedAuthMo.getString(KEY_IDENTITY);
            return new PresharedAuthenticationData(identity);
        } catch (e) {
            if (e instanceof MslEncoderException)
                throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "psk authdata" + presharedAuthMo);
            throw e;
        }
    };

    // Exports.
    module.exports.parse = PresharedAuthenticationData$parse;
})(require, (typeof module !== 'undefined') ? module : mkmodule('PresharedAuthenticationData'));
